### ISSUE TITLE
[Internal tag reader] Sort by Year bug

### ISSUE THREAD

**Issue by andiandi13** [vanilla-music/vanilla#625]
- Vanilla Music Version: 1.0.53
- Android Version: 4.4.4
- Phone vendor / model: Samsung Galaxy S5 (SM-G901F)
- Application was downloaded from:
  - [ ] Play Store
  - [x] F-Droid
  - [x] Other (Nightly Build)
- Did this work in a previous version?
  - [ ] Yes
  - [ ] No
  - [x] NA / Don't know.

## Description

The sorting by Year when  "Always use internal tag reader" is checked behaves so strangely.

It sort by Year (From 2016 to 1986) but after 1986 (being my smallest number except year 0 tag) I _can_ continue and there is songs that are sorted by **Album Name** :open_mouth:

If I uncheck the internal tag reader box, I can go until Year 0 without any issues.

To sum up it looks like that :

```
Play all
2016
2015
2010
2008
1990
1986
Z
Y
X
T
R
P
G
D
C
B
A
9
```

## Steps to reproduce the issue

Check internal tag reader in media library screen, then do a full scan.

## Videos

Please check those : 

With Internal Tag reader : [YouTube](https://youtu.be/oHygYrDRbNI)
Without Internal Tag reader : [YouTube](https://youtu.be/myUkNeLsrfc)

**Comment by kirbyfan64** [comment]
I smell an epoch-related issue...

**Comment by adrian-bl** [comment]
I assume you meant that you CANT continue?

Thats probably because the internal tagreader wont report 0 as a year - and the fast scroller will get a NULL value which will not be displayed

**Comment by andiandi13** [comment]
No, I really mean that I can continue as you can see on the video.

In fact, the songs above 1986 are from different years (2015, 2009, 2004, 2005, 2008 etc.. AND 0 but that's not the problem)

Instead of sorting all those years 
that I just mentioned above, within the others, it just separate them from the "Year" sorting list and put them on a second "Album" list.

There is two different sorting when I choose Year..

**Comment by adrian-bl** [comment]
> No, I really mean that I can continue as you can see on the video.

What i meant was that you can't continue using the fastscroller (but you can continue using the normal scroll)

This should be fixed by https://github.com/vanilla-music/vanilla/commit/0d4c8685ae993896cff566be8023c51868bae405 - we will now display '?' in the fastscroller if we get a NULL value (instead of omitting it).

> In fact, the songs above 1986 are from different years (2015, 2009, 2004, 2005, 2008 etc.. AND 0 but 
> that's not the problem)

So you are saying that songs 'above' 1986 are tagged with a year?
It seems that we just were not able to read the tag correctly in this case and indexed the file with year=NULL - feel free to send me a sample to adrian@blinkenlights.ch

**Comment by andiandi13** [comment]
> What i meant was that you can't continue using the fastscroller (but you can continue using the normal scroll)

Yes its exactly like that.

> So you are saying that songs 'above' 1986 are tagged with a year?

Not every songs, but I think I figured out a thing: Above 1986 there is only mp3 files that are tagged with Year '0' AND .m4a files, even if they are tags with a Year. 

So it seems that there is a problem with m4a tags, I just sent you a m4a from 2013 that is placed above 1986, please tell me if it's above your "oldest" song.

Thanks

**Comment by andiandi13** [comment]
EDIT : commit #0d4c868 works great in the latest build. We just have to figure out the bad sorting of m4a years.

**Comment by adrian-bl** [comment]
Our m4a parser is a little bit shaky, but commit https://github.com/adrian-bl/bastp/commit/9c8eda9134881613b19a7ea3f62f71ea5365f3da seems to fix the issue:

```
find t/Post* | java ch.blinkenlights.bastp.Test

------------------------------------------------------------
<FILE> t/Post Scriptum - Kery James.m4a
<ALBUM> = [Dernier Mc]
<YEAR> = [2013]
<ARTIST> = [Kery James]
<type> = MP4
<TITLE> = [Post Scriptum - Kery James]
== TEST RESULTS, PARSED 1 FILE(S)
== BROKEN = 0, FLAC = 0, OGG = 0, ID3v2 = 0, OTHER = 1, TAGS = 5
```

**Comment by andiandi13** [comment]
> Our m4a parser is a little bit shaky, but commit adrian-bl/bastp@9c8eda9 seems to fix the issue

Alright thank you, I'll wait or the next nightly to try it.

**Comment by adrian-bl** [comment]
The current nightly already includes the fix (i triggered a rebuild) - note that this only affects newly indexed songs: So you either have to remove + copy the songs back to the device or run a full scan (with a database wipe)

**Comment by andiandi13** [comment]
It works perfectly after a rescan, thanks again :)
